File: src/main/java/com/upokecenter/util/DataIO.java
package com.upokecenter.util;
/*
Written by Peter O.
Any copyright to this work is released to the Public Domain.
In case this is not possible, this work is also
licensed under the Unlicense: https://unlicense.org/

 */

import java.io.*;

  /**
   * Convenience class that contains static methods for wrapping byte arrays and
   * streams into byte readers and byte writers.
   */
  public final class DataIO {
private DataIO() {
}
    /**
     * <p>Wraps a byte array into a byte reader. The reader will start at the
     * beginning of the byte array. </p> <p>In the.NET implementation, this method
     * is implemented as an extension method to any byte array object and can be
     * called as follows: {@code bytes.ToByteReader()}. If the object's class
     * already has a ToByteReader method with the same parameters, that method
     * takes precedence over this extension method.</p>
     * @param bytes The byte array to wrap.
     * @return A byte reader wrapping the byte array.
     * @throws NullPointerException The parameter {@code bytes} is null.
     */
    public static IReader ToReader(
       byte[] bytes) {
      if (bytes == null) {
        throw new NullPointerException("bytes");
      }
      return new ByteArrayTransform(bytes, 0, bytes.length);
    }

    /**
     * <p>Wraps a portion of a byte array into a byte reader object. </p> <p>In
     * the.NET implementation, this method is implemented as an extension method to
     * any byte array object and can be called as follows: {@code
     * bytes.ToByteReader(offset, length)}. If the object's class already has a
     * ToByteReader method with the same parameters, that method takes precedence
     * over this extension method.</p>
     * @param bytes The byte array to wrap.
     * @param offset An index starting at 0 showing where the desired portion of
     * "bytes" begins.
     * @param length The length, in bytes, of the desired portion of "bytes" (but
     * not more than "bytes" 's length).
     * @return A byte reader wrapping the byte array.
     * @throws NullPointerException The parameter {@code bytes} is null.
     * @throws IllegalArgumentException Either {@code offset} or {@code length} is less
     * than 0 or greater than {@code bytes} 's length, or {@code bytes} 's length
     * minus {@code offset} is less than {@code length}.
     * @throws NullPointerException The parameter {@code bytes} is null.
     */
    public static IReader ToReader(
      byte[] bytes,
      int offset,
      int length) {
      if (bytes == null) {
        throw new NullPointerException("bytes");
      }
      if (offset < 0) {
        throw new IllegalArgumentException("offset(" + offset +
          ") is less than 0");
      }
      if (offset > bytes.length) {
        throw new IllegalArgumentException("offset(" + offset + ") is more than " +
          bytes.length);
      }
      if (length < 0) {
        throw new IllegalArgumentException("length(" + length +
          ") is less than 0");
      }
      if (length > bytes.length) {
        throw new IllegalArgumentException("length(" + length + ") is more than " +
          bytes.length);
      }
      if (bytes.length - offset < length) {
        throw new IllegalArgumentException("bytes's length minus " + offset + "(" +
          (bytes.length - offset) + ") is less than " + length);
      }
      return new ByteArrayTransform(bytes, offset, length);
    }

    /**
     * <p>Wraps an input stream into a reader object. If an IOException is thrown
     * by the input stream, the reader object throws IllegalStateException instead.
     * </p> <p>In the.NET implementation, this method is implemented as an
     * extension method to any object implementing InputStream and can be called as
     * follows: {@code input.ToByteReader()}. If the object's class already has a
     * ToByteReader method with the same parameters, that method takes precedence
     * over this extension method.</p>
     * @param input The input stream to wrap.
     * @return A byte reader wrapping the input stream.
     * @throws NullPointerException The parameter {@code input} is null.
     */
    public static IReader ToReader(
       InputStream input) {
      if (input == null) {
        throw new NullPointerException("input");
      }
      return new WrappedStream(input);
    }

    /**
     * <p>Wraps a portion of a byte array into a byte reader. </p> <p>In the.NET
     * implementation, this method is implemented as an extension method to any
     * object implementing byte[] and can be called as follows: {@code
     * bytes.ToByteReader(offset, length)}. If the object's class already has a
     * ToByteReader method with the same parameters, that method takes precedence
     * over this extension method.</p>
     * @param bytes The byte array to wrap into a byte reader.
     * @param offset An index starting at 0 showing where the desired portion of
     * {@code bytes} begins.
     * @param length The length, in bytes, of the desired portion of {@code bytes}
     * (but not more than {@code bytes} 's length).
     * @return An IByteReader object.
     * @throws NullPointerException The parameter {@code bytes} is null.
     * @throws IllegalArgumentException Either {@code offset} or {@code length} is less
     * than 0 or greater than {@code bytes} 's length, or {@code bytes} 's length
     * minus {@code offset} is less than {@code length}.
     * @throws NullPointerException The parameter {@code bytes} is null.
     * @deprecated Use ToReader instead.
 */
@Deprecated
    public static IByteReader ToByteReader(
      byte[] bytes,
      int offset,
      int length) {
      return (IByteReader)ToReader(bytes, offset, length);
    }

    /**
     * <p>Wraps a data stream into a byte reader. </p> <p>In the.NET
     * implementation, this method is implemented as an extension method to any
     * object implementing InputStream and can be called as follows: {@code
     * input.ToByteReader()}. If the object's class already has a ToByteReader
     * method with the same parameters, that method takes precedence over this
     * extension method.</p>
     * @param input The data stream to wrap into a byte reader.
     * @return An IByteReader object.
     * @throws NullPointerException The parameter {@code input} is null.
     * @deprecated Use ToReader instead.
 */
@Deprecated
    public static IByteReader ToByteReader(
       InputStream input) {
      return (IByteReader)ToReader(input);
    }

    /**
     * <p>Wraps a byte array into a byte reader. </p> <p>In the.NET implementation,
     * this method is implemented as an extension method to any object implementing
     * byte[] and can be called as follows: {@code bytes.ToByteReader()}. If the
     * object's class already has a ToByteReader method with the same parameters,
     * that method takes precedence over this extension method.</p>
     * @param bytes The byte array to wrap into a byte reader.
     * @return An IByteReader object.
     * @throws NullPointerException The parameter {@code bytes} is null.
     * @deprecated Use ToReader instead.
 */
@Deprecated
    public static IByteReader ToByteReader(
       byte[] bytes) {
      return (IByteReader)ToReader(bytes);
    }

    /**
     * <p>Wraps an output stream into a writer object. If an IOException is thrown
     * by the input stream, the writer object throws IllegalStateException instead.
     * </p> <p>In the.NET implementation, this method is implemented as an
     * extension method to any object implementing InputStream and can be called as
     * follows: {@code output.ToWriter()}. If the object's class already has a
     * ToWriter method with the same parameters, that method takes precedence over
     * this extension method.</p>
     * @param output Output stream to wrap.
     * @return A byte writer that wraps the specified output stream.
     * @throws NullPointerException The parameter {@code output} is null.
     */
    public static IWriter ToWriter(
       OutputStream output) {
      if (output == null) {
        throw new NullPointerException("output");
      }
      return new WrappedOutputStream(output);
    }

    /**
     * <p>Wraps a byte writer (one that only implements a ReadByte method) to a
     * writer (one that also implements a three-parameter Read method.) </p> <p>In
     * the.NET implementation, this method is implemented as an extension method to
     * any object implementing IByteWriter and can be called as follows: {@code
     * output.ToWriter()}. If the object's class already has a ToWriter method with
     * the same parameters, that method takes precedence over this extension
     * method.</p>
     * @param output A byte stream.
     * @return A writer that wraps the specified stream.
     * @throws NullPointerException The parameter {@code output} is null.
     */
    public static IWriter ToWriter(
       IByteWriter output) {
      if (output == null) {
        throw new NullPointerException("output");
      }
      return new WrappedOutputStreamFromByteWriter(output);
    }

    private static final class ByteArrayTransform implements IReader {
      private final byte[] bytes;
      private final int endOffset;
      private int offset;

      public ByteArrayTransform(byte[] bytes, int offset, int length) {
        this.bytes = bytes;
        this.offset = offset;
        this.endOffset = offset + length;
      }

      /**
       * This is an internal method.
       * @return A 32-bit signed integer.
       */
      public int read() {
        if (this.offset >= this.endOffset) {
          return -1;
        }
        int b = this.bytes[this.offset];
        ++this.offset;
        return ((int)b) & 0xff;
      }

      /**
       * This is an internal method.
       * @param bytes The parameter {@code bytes} is an internal parameter.
       * @param offset An index starting at 0 showing where the desired portion of
       * {@code bytes} begins.
       * @param length The length, in bytes, of the desired portion of {@code bytes}
       * (but not more than {@code bytes} 's length).
       * @return A 32-bit signed integer.
       * @throws NullPointerException The parameter {@code bytes} is null.
       * @throws IllegalArgumentException Either {@code offset} or {@code length} is less
       * than 0 or greater than {@code bytes} 's length, or {@code bytes} 's length
       * minus {@code offset} is less than {@code length}.
       */
      public int Read(byte[] bytes, int offset, int length) {
        if (bytes == null) {
          throw new NullPointerException("bytes");
        }
        if (offset < 0) {
          throw new IllegalArgumentException("offset(" + offset +
            ") is less than 0");
        }
        if (offset > bytes.length) {
          throw new IllegalArgumentException("offset(" + offset +
            ") is more than " + bytes.length);
        }
        if (length < 0) {
          throw new IllegalArgumentException("length(" + length +
            ") is less than 0");
        }
        if (length > bytes.length) {
          throw new IllegalArgumentException("length(" + length +
            ") is more than " + bytes.length);
        }
        if (bytes.length - offset < length) {
          throw new IllegalArgumentException("bytes's length minus " + offset + "(" +
            (bytes.length - offset) + ") is less than " + length);
        }
        int count = 0;
        for (int i = 0; i < length; ++i) {
          int c = this.read();
          if (c == -1) {
            break;
          }
          bytes[offset] = (byte)(c & 0xff);
          ++count;
          ++offset;
        }
        return count;
      }
    }

    private static final class WrappedOutputStream implements IWriter {
      private final OutputStream output;

      public WrappedOutputStream(OutputStream output) {
        this.output = output;
      }

      /**
       * This is an internal method.
       * @param byteValue The parameter {@code byteValue} is a 32-bit signed integer.
       */
      public void write(int byteValue) {
        try {
          this.output.write((byte)byteValue);
        } catch (IOException ex) {
          throw new IllegalStateException(ex.getMessage(), ex);
        }
      }

      /**
       * This is an internal method.
       * @param bytes A byte array.
       * @param offset An index starting at 0 showing where the desired portion of
       * "bytes" begins.
       * @param length The length, in bytes, of the desired portion of "bytes" (but
       * not more than "bytes" 's length).
       * @throws NullPointerException The parameter {@code bytes} is null.
       * @throws IllegalArgumentException Either {@code offset} or {@code length} is less
       * than 0 or greater than {@code bytes} 's length, or {@code bytes} 's length
       * minus {@code offset} is less than {@code length}.
       */
      public void write(byte[] bytes, int offset, int length) {
        try {
          this.output.write(bytes, offset, length);
        } catch (IOException ex) {
          throw new IllegalStateException(ex.getMessage(), ex);
        }
      }
    }

    private static final class WrappedOutputStreamFromByteWriter implements IWriter {
      private final IByteWriter output;

      public WrappedOutputStreamFromByteWriter(IByteWriter output) {
        this.output = output;
      }

      /**
       * This is an internal method.
       * @param byteValue The parameter {@code byteValue} is a 32-bit signed integer.
       */
      public void write(int byteValue) {
        this.output.write((byte)byteValue);
      }

      /**
       * This is an internal method.
       * @param bytes A byte array.
       * @param offset An index starting at 0 showing where the desired portion of
       * "bytes" begins.
       * @param length The length, in bytes, of the desired portion of "bytes" (but
       * not more than "bytes" 's length).
       * @throws NullPointerException The parameter {@code bytes} is null.
       * @throws IllegalArgumentException Either {@code offset} or {@code length} is less
       * than 0 or greater than {@code bytes} 's length, or {@code bytes} 's length
       * minus {@code offset} is less than {@code length}.
       * @throws NullPointerException The parameter {@code bytes} is null.
       */
      public void write(byte[] bytes, int offset, int length) {
        if (bytes == null) {
          throw new NullPointerException("bytes");
        }
        if (offset < 0) {
          throw new IllegalArgumentException("offset(" + offset +
            ") is less than 0");
        }
        if (offset > bytes.length) {
          throw new IllegalArgumentException("offset(" + offset + ") is more than " +
            bytes.length);
        }
        if (length < 0) {
          throw new IllegalArgumentException("length(" + length +
            ") is less than 0");
        }
        if (length > bytes.length) {
          throw new IllegalArgumentException("length(" + length + ") is more than " +
            bytes.length);
        }
        if (bytes.length - offset < length) {
          throw new IllegalArgumentException("bytes's length minus " + offset + "(" +
            (bytes.length - offset) + ") is less than " + length);
        }
        for (int i = 0; i < length; ++i) {
          this.output.write((byte)bytes[i]);
        }
      }
    }

    private static final class WrappedStream implements IReader {
      private final InputStream stream;

      public WrappedStream(InputStream stream) {
        this.stream = stream;
      }

      /**
       * This is an internal method.
       * @return A 32-bit signed integer.
       */
      public int read() {
        try {
          return this.stream.read();
        } catch (IOException ex) {
          throw new IllegalStateException(ex.getMessage(), ex);
        }
      }

      public int Read(byte[] bytes, int offset, int length) {
        try {
          return Math.max(0, this.stream.read(bytes, offset, length));
        } catch (IOException ex) {
          throw new IllegalStateException(ex.getMessage(), ex);
        }
      }
    }
  }

File: Encoding/PeterO/DataIO.cs
/*
Written by Peter O.
Any copyright to this work is released to the Public Domain.
In case this is not possible, this work is also
licensed under the Unlicense: https://unlicense.org/

 */
using System;
using System.IO;

namespace PeterO {
  /// <summary>Convenience class that contains static methods for
  /// wrapping byte arrays and streams into byte readers and byte
  /// writers.</summary>
  public static class DataIO {
    /// <summary>Wraps a byte array into a byte reader. The reader will
    /// start at the beginning of the byte array.
    /// <para>In the.NET implementation, this method is implemented as an
    /// extension method to any byte array object and can be called as
    /// follows: <c>bytes.ToByteReader()</c>. If the object's class
    /// already has a ToByteReader method with the same parameters, that
    /// method takes precedence over this extension
    /// method.</para></summary>
    /// <param name='bytes'>The byte array to wrap.</param>
    /// <returns>A byte reader wrapping the byte array.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='bytes'/> is null.</exception>
    public static IReader ToReader(
      #if !NET20
       this
 #endif

      byte[] bytes) {
      if (bytes == null) {
        throw new ArgumentNullException(nameof(bytes));
      }
      return new ByteArrayTransform(bytes, 0, bytes.Length);
    }

    /// <summary>Wraps a portion of a byte array into a byte reader object.
    /// <para>In the.NET implementation, this method is implemented as an
    /// extension method to any byte array object and can be called as
    /// follows: <c>bytes.ToByteReader(offset, length)</c>. If the
    /// object's class already has a ToByteReader method with the same
    /// parameters, that method takes precedence over this extension
    /// method.</para></summary>
    /// <param name='bytes'>The byte array to wrap.</param>
    /// <param name='offset'>An index starting at 0 showing where the
    /// desired portion of "bytes" begins.</param>
    /// <param name='length'>The length, in bytes, of the desired portion
    /// of "bytes" (but not more than "bytes" 's length).</param>
    /// <returns>A byte reader wrapping the byte array.</returns>
    /// <exception cref=' T:System.ArgumentNullException'>The parameter
    /// <paramref name='bytes'/> is null.</exception>
    /// <exception cref='ArgumentException'>Either <paramref
    /// name='offset'/> or <paramref name='length'/> is less than 0 or
    /// greater than <paramref name='bytes'/> 's length, or <paramref
    /// name=' bytes'/> 's length minus <paramref name='offset'/> is less
    /// than <paramref name='length'/>.</exception>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='bytes'/> is null.</exception>
    public static IReader ToReader(
      #if !NET20
      this
      #endif
      byte[] bytes,
      int offset,
      int length) {
      if (bytes == null) {
        throw new ArgumentNullException(nameof(bytes));
      }
      if (offset < 0) {
        throw new ArgumentException("offset(" + offset +
          ") is less than 0");
      }
      if (offset > bytes.Length) {
        throw new ArgumentException("offset(" + offset + ") is more than " +
          bytes.Length);
      }
      if (length < 0) {
        throw new ArgumentException("length(" + length +
          ") is less than 0");
      }
      if (length > bytes.Length) {
        throw new ArgumentException("length(" + length + ") is more than " +
          bytes.Length);
      }
      if (bytes.Length - offset < length) {
        throw new ArgumentException("bytes's length minus " + offset + "(" +
          (bytes.Length - offset) + ") is less than " + length);
      }
      return new ByteArrayTransform(bytes, offset, length);
    }

    /// <summary>Wraps an input stream into a reader object. If an
    /// IOException is thrown by the input stream, the reader object throws
    /// InvalidOperationException instead.
    /// <para>In the.NET implementation, this method is implemented as an
    /// extension method to any object implementing Stream and can be
    /// called as follows: <c>input.ToByteReader()</c>. If the object's
    /// class already has a ToByteReader method with the same parameters,
    /// that method takes precedence over this extension
    /// method.</para></summary>
    /// <param name='input'>The input stream to wrap.</param>
    /// <returns>A byte reader wrapping the input stream.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='input'/> is null.</exception>
    public static IReader ToReader(
      #if !NET20
       this
 #endif

      Stream input) {
      if (input == null) {
        throw new ArgumentNullException(nameof(input));
      }
      return new WrappedStream(input);
    }

    /// <summary>Wraps a portion of a byte array into a byte reader.
    /// <para>In the.NET implementation, this method is implemented as an
    /// extension method to any object implementing byte[] and can be
    /// called as follows: <c>bytes.ToByteReader(offset, length)</c>. If
    /// the object's class already has a ToByteReader method with the same
    /// parameters, that method takes precedence over this extension
    /// method.</para></summary>
    /// <param name='bytes'>The byte array to wrap into a byte
    /// reader.</param>
    /// <param name='offset'>An index starting at 0 showing where the
    /// desired portion of <paramref name='bytes'/> begins.</param>
    /// <param name='length'>The length, in bytes, of the desired portion
    /// of <paramref name='bytes'/> (but not more than <paramref
    /// name='bytes'/> 's length).</param>
    /// <returns>An IByteReader object.</returns>
    /// <exception cref=' T:System.ArgumentNullException'>The parameter
    /// <paramref name='bytes'/> is null.</exception>
    /// <exception cref='ArgumentException'>Either <paramref
    /// name='offset'/> or <paramref name='length'/> is less than 0 or
    /// greater than <paramref name='bytes'/> 's length, or <paramref
    /// name=' bytes'/> 's length minus <paramref name='offset'/> is less
    /// than <paramref name='length'/>.</exception>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='bytes'/> is null.</exception>
    [Obsolete("Use ToReader instead.")]
    public static IByteReader ToByteReader(
      #if !NET20
      this
      #endif
      byte[] bytes,
      int offset,
      int length) {
      return (IByteReader)ToReader(bytes, offset, length);
    }

    /// <summary>Wraps a data stream into a byte reader.
    /// <para>In the.NET implementation, this method is implemented as an
    /// extension method to any object implementing Stream and can be
    /// called as follows: <c>input.ToByteReader()</c>. If the object's
    /// class already has a ToByteReader method with the same parameters,
    /// that method takes precedence over this extension
    /// method.</para></summary>
    /// <param name='input'>The data stream to wrap into a byte
    /// reader.</param>
    /// <returns>An IByteReader object.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='input'/> is null.</exception>
    [Obsolete("Use ToReader instead.")]
    public static IByteReader ToByteReader(
      #if !NET20
       this
 #endif

      Stream input) {
      return (IByteReader)ToReader(input);
    }

    /// <summary>Wraps a byte array into a byte reader.
    /// <para>In the.NET implementation, this method is implemented as an
    /// extension method to any object implementing byte[] and can be
    /// called as follows: <c>bytes.ToByteReader()</c>. If the object's
    /// class already has a ToByteReader method with the same parameters,
    /// that method takes precedence over this extension
    /// method.</para></summary>
    /// <param name='bytes'>The byte array to wrap into a byte
    /// reader.</param>
    /// <returns>An IByteReader object.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='bytes'/> is null.</exception>
    [Obsolete("Use ToReader instead.")]
    public static IByteReader ToByteReader(
      #if !NET20
       this
 #endif

      byte[] bytes) {
      return (IByteReader)ToReader(bytes);
    }

    /// <summary>Wraps an output stream into a writer object. If an
    /// IOException is thrown by the input stream, the writer object throws
    /// InvalidOperationException instead.
    /// <para>In the.NET implementation, this method is implemented as an
    /// extension method to any object implementing Stream and can be
    /// called as follows: <c>output.ToWriter()</c>. If the object's class
    /// already has a ToWriter method with the same parameters, that method
    /// takes precedence over this extension method.</para></summary>
    /// <param name='output'>Output stream to wrap.</param>
    /// <returns>A byte writer that wraps the specified output
    /// stream.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='output'/> is null.</exception>
    public static IWriter ToWriter(
      #if !NET20
       this
 #endif

      Stream output) {
      if (output == null) {
        throw new ArgumentNullException(nameof(output));
      }
      return new WrappedOutputStream(output);
    }

    /// <summary>Wraps a byte writer (one that only implements a ReadByte
    /// method) to a writer (one that also implements a three-parameter
    /// Read method.)
    /// <para>In the.NET implementation, this method is implemented as an
    /// extension method to any object implementing IByteWriter and can be
    /// called as follows: <c>output.ToWriter()</c>. If the object's class
    /// already has a ToWriter method with the same parameters, that method
    /// takes precedence over this extension method.</para></summary>
    /// <param name='output'>A byte stream.</param>
    /// <returns>A writer that wraps the specified stream.</returns>
    /// <exception cref='ArgumentNullException'>The parameter <paramref
    /// name='output'/> is null.</exception>
    public static IWriter ToWriter(
      #if !NET20
       this
 #endif

      IByteWriter output) {
      if (output == null) {
        throw new ArgumentNullException(nameof(output));
      }
      return new WrappedOutputStreamFromByteWriter(output);
    }

    private sealed class ByteArrayTransform : IReader {
      private readonly byte[] bytes;
      private readonly int endOffset;
      private int offset;

      public ByteArrayTransform(byte[] bytes, int offset, int length) {
        this.bytes = bytes;
        this.offset = offset;
        this.endOffset = offset + length;
      }

      /// <summary>This is an internal method.</summary>
      /// <returns>A 32-bit signed integer.</returns>
      public int ReadByte() {
        if (this.offset >= this.endOffset) {
          return -1;
        }
        int b = this.bytes[this.offset];
        ++this.offset;
        return ((int)b) & 0xff;
      }

      /// <summary>This is an internal method.</summary>
      /// <param name='bytes'>The parameter <paramref name='bytes'/> is an
      /// internal parameter.</param>
      /// <param name='offset'>An index starting at 0 showing where the
      /// desired portion of <paramref name='bytes'/> begins.</param>
      /// <param name='length'>The length, in bytes, of the desired portion
      /// of <paramref name='bytes'/> (but not more than <paramref
      /// name='bytes'/> 's length).</param>
      /// <returns>A 32-bit signed integer.</returns>
      /// <exception cref='ArgumentNullException'>The parameter <paramref
      /// name='bytes'/> is null.</exception>
      /// <exception cref='ArgumentException'>Either <paramref
      /// name='offset'/> or <paramref name='length'/> is less than 0 or
      /// greater than <paramref name='bytes'/> 's length, or <paramref
      /// name='bytes'/> 's length minus <paramref name='offset'/> is less
      /// than <paramref name='length'/>.</exception>
      public int Read(byte[] bytes, int offset, int length) {
        if (bytes == null) {
          throw new ArgumentNullException(nameof(bytes));
        }
        if (offset < 0) {
          throw new ArgumentException("offset(" + offset +
            ") is less than 0");
        }
        if (offset > bytes.Length) {
          throw new ArgumentException("offset(" + offset +
            ") is more than " + bytes.Length);
        }
        if (length < 0) {
          throw new ArgumentException("length(" + length +
            ") is less than 0");
        }
        if (length > bytes.Length) {
          throw new ArgumentException("length(" + length +
            ") is more than " + bytes.Length);
        }
        if (bytes.Length - offset < length) {
          throw new ArgumentException("bytes's length minus " + offset + "(" +
            (bytes.Length - offset) + ") is less than " + length);
        }
        var count = 0;
        for (var i = 0; i < length; ++i) {
          int c = this.ReadByte();
          if (c == -1) {
            break;
          }
          bytes[offset] = (byte)(c & 0xff);
          ++count;
          ++offset;
        }
        return count;
      }
    }

    private sealed class WrappedOutputStream : IWriter {
      private readonly Stream output;

      public WrappedOutputStream(Stream output) {
        this.output = output;
      }

      /// <summary>This is an internal method.</summary>
      /// <param name='byteValue'>The parameter <paramref name='byteValue'/>
      /// is a 32-bit signed integer.</param>
      public void WriteByte(int byteValue) {
        try {
          this.output.WriteByte((byte)byteValue);
        } catch (IOException ex) {
          throw new InvalidOperationException(ex.Message, ex);
        }
      }

      /// <summary>This is an internal method.</summary>
      /// <param name='bytes'>A byte array.</param>
      /// <param name='offset'>An index starting at 0 showing where the
      /// desired portion of "bytes" begins.</param>
      /// <param name='length'>The length, in bytes, of the desired portion
      /// of "bytes" (but not more than "bytes" 's length).</param>
      /// <exception cref='ArgumentNullException'>The parameter <paramref
      /// name='bytes'/> is null.</exception>
      /// <exception cref='ArgumentException'>Either <paramref
      /// name='offset'/> or <paramref name='length'/> is less than 0 or
      /// greater than <paramref name='bytes'/> 's length, or <paramref
      /// name='bytes'/> 's length minus <paramref name='offset'/> is less
      /// than <paramref name='length'/>.</exception>
      public void Write(byte[] bytes, int offset, int length) {
        try {
          this.output.Write(bytes, offset, length);
        } catch (IOException ex) {
          throw new InvalidOperationException(ex.Message, ex);
        }
      }
    }

    private sealed class WrappedOutputStreamFromByteWriter : IWriter {
      private readonly IByteWriter output;

      public WrappedOutputStreamFromByteWriter(IByteWriter output) {
        this.output = output;
      }

      /// <summary>This is an internal method.</summary>
      /// <param name='byteValue'>The parameter <paramref name='byteValue'/>
      /// is a 32-bit signed integer.</param>
      public void WriteByte(int byteValue) {
        this.output.WriteByte((byte)byteValue);
      }

      /// <summary>This is an internal method.</summary>
      /// <param name='bytes'>A byte array.</param>
      /// <param name='offset'>An index starting at 0 showing where the
      /// desired portion of "bytes" begins.</param>
      /// <param name='length'>The length, in bytes, of the desired portion
      /// of "bytes" (but not more than "bytes" 's length).</param>
      /// <exception cref=' T:System.ArgumentNullException'>The parameter
      /// <paramref name='bytes'/> is null.</exception>
      /// <exception cref='ArgumentException'>Either <paramref
      /// name='offset'/> or <paramref name='length'/> is less than 0 or
      /// greater than <paramref name='bytes'/> 's length, or <paramref
      /// name=' bytes'/> 's length minus <paramref name='offset'/> is less
      /// than <paramref name='length'/>.</exception>
      /// <exception cref='ArgumentNullException'>The parameter <paramref
      /// name='bytes'/> is null.</exception>
      public void Write(byte[] bytes, int offset, int length) {
        if (bytes == null) {
          throw new ArgumentNullException(nameof(bytes));
        }
        if (offset < 0) {
          throw new ArgumentException("offset(" + offset +
            ") is less than 0");
        }
        if (offset > bytes.Length) {
          throw new ArgumentException("offset(" + offset + ") is more than " +
            bytes.Length);
        }
        if (length < 0) {
          throw new ArgumentException("length(" + length +
            ") is less than 0");
        }
        if (length > bytes.Length) {
          throw new ArgumentException("length(" + length + ") is more than " +
            bytes.Length);
        }
        if (bytes.Length - offset < length) {
          throw new ArgumentException("bytes's length minus " + offset + "(" +
            (bytes.Length - offset) + ") is less than " + length);
        }
        for (int i = 0; i < length; ++i) {
          this.output.WriteByte((byte)bytes[i]);
        }
      }
    }

    private sealed class WrappedStream : IReader {
      private readonly Stream stream;

      public WrappedStream(Stream stream) {
        this.stream = stream;
      }

      /// <summary>This is an internal method.</summary>
      /// <returns>A 32-bit signed integer.</returns>
      public int ReadByte() {
        try {
          return this.stream.ReadByte();
        } catch (IOException ex) {
          throw new InvalidOperationException(ex.Message, ex);
        }
      }

      public int Read(byte[] bytes, int offset, int length) {
        try {
          return Math.Max(0, this.stream.Read(bytes, offset, length));
        } catch (IOException ex) {
          throw new InvalidOperationException(ex.Message, ex);
        }
      }
    }
  }
}
